only use YAML parser for YAML responses

arty-name opened the pull request:
### Description
Check the content-type of response and use appropriate parser: `js-yaml`, json, or no parser.

### Motivation and Context
As described in #1165 even if one doesn’t use YAML at all, providing the spec via JSON object and using JSON to talk to backend, the swagger-client still uses `js-yaml` to parse responses. The downside to that approach is that one cannot forcefully remove `js-yaml` from Webpack bundle. 

If someone however have been using invalid JSON in the responses, this change will break things. Apparently the JSON parser of `js-yaml` is quite relaxed and handles even invalid JSON like this `{"123": NULL}`. It is your call whether to treat this as a breaking change or not.

### How Has This Been Tested?
`npm test` has passed

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] No code changes (changes to documentation, CI, metadata, etc)
- [ ] Dependency changes (any modification to dependencies in `package.json`)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

shockey commented:
@arty-name, my suggestion going forward here would be to do a naive content check, something like:

```
if(req.body.trim()[0] === "{") {
  // json 
} else { 
  // yaml
 }
```

---

Also, concerning your use case of wanting to jettison `js-yaml`: have you considered aliasing `js-yaml` to `JSON.parse`? Like this:

```js
// alias.js: `js-yaml` interface mock

export default {
  safeLoad(json) {
    return JSON.parse(json)
  }
}
```

```js
// webpack.config.js
resolve: { 
  alias: { 
    "js-yaml": path.resolve(__dirname, 'alias.js') 
  }
}
```

arty-name commented:
Thank you for adding tests, @shockey! I have updated my change as you suggest - almost. For performance reasons I have replaced trimming with regexp. How does it look now?

The tip with aliasing safeLoad to JSON.parse will technically work, but it looks to me too much like a tight coupling.

shockey commented:
Thanks @arty-name!

arty-name commented:
@shockey, there’s actually something wrong about the OpenAPI argument. 

Yes, the OpenAPI Document is always an object. However the specification allows the **responses** to be arrays, it even includes [examples](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#responseObject). 

The code in `serializeRes` is used for both Swagger specs and responses, so it’s perfectly fine for it to receive a JSON-encoded array, string, number, null, or object. Naive content check seems not too reliable for that case.

shockey commented:
@arty-name, you're correct.

Fortunately, this shouldn't create any issues for folks that aren't excluding `js-yaml` from their projects 😄 

I'd be open to a second swing on this, where we swap out the naive check for `contentType === 'application/json'`. 

This would require you (and others that want to ditch `js-yaml`) to always be sure to declare a content type, but would handle the issue you've pointed out.

arty-name commented:
I’m happy to hear that you don’t mind to check another pull request for this issue! I have created #1183 with this approach.
